Full BIP173 (Bech32) support

sipa opened the pull request:
Builds on top of #11117.

This adds support for:
* Creating BIP173 addresses for testing (through `addwitnessaddress`, though by default it still produces P2SH versions)
* Sending to BIP173 addresses (including non-v0 ones)
* Analysing BIP173 addresses (through `validateaddress`)

It includes a reformatted version of the [C++ Bech32 reference code](https://github.com/sipa/bech32/tree/master/ref/c%2B%2B) and an independent implementation of the address encoding/decoding logic (integrated with CTxDestination). All BIP173 test vectors are included.

Not included (and intended for other PRs):
* Full wallet support for SegWit (which would include automatically adding witness scripts to the wallet during automatic keypool topup, SegWit change outputs, ...) [see #11403]
* Splitting base58.cpp and tests/base58_tests.cpp up into base58-specific code, and "address encoding"-code [see #11372]
* Error locating in UI for BIP173 addresses.

sipa committed:
Implement {Encode,Decode}Destination without CBitcoinAddress

sipa committed:
Import Bech32 C++ reference code & tests

This includes a reformatted version of the Bech32 reference code
(see https://github.com/sipa/bech32/tree/master/ref/c%2B%2B), with
extra documentation.

sipa committed:
Implement BIP173 addresses and tests

sipa committed:
Support BIP173 in addwitnessaddress

luke-jr commented:
Would prefer to have simply sending-to (maybe validating/analyzing too?) as a separate PR, before wallet upgrades.

I'm not sure when it would make sense to convert between P2SH and BIP173...

sipa commented:
@luke-jr I agree, but I consider `addwitnessaddress` an RPC to aid with testing, not full support.

gmaxwell commented:
addwitnessaddress is very much not actual support, it's a test shim.

luke-jr commented:
But it modifies the wallet, no? Seems useful to review independently from the rest. Especially since it has the additional considerations of what happens if you try to use it and then downgrade to an older version...

sipa commented:
@luke-jr It does. But consider that we've since 0.13.1 had support to receiving and spending native witness outputs in the wallet (without that, testing the consensus logic for it would have been much harder), just no way to encode such outputs as strings. This merely extends that ability to test to include the string representation.

sipa committed:
Use BIP173 addresses in segwit.py test

sipa commented:
Added support in Python framework, and some integrated some functional tests into the `segwit.py` test.

promag commented on src/bech32.cpp:
```diff
@@ -0,0 +1,157 @@
+}
+
+/** Decode a Bech32 string. */
+std::pair<std::string, data> Decode(const std::string& str) {
```
IMO `bool Decode(const std::string& str, const std::string& hrp, data& d)` feels better, and this way below it can early return.

sipa replied:
It's not often used in the Bitcoin Core codebase, but using pairs for multiple returned values is very typical in C++ (see the return type of [std::map::insert](http://en.cppreference.com/w/cpp/container/map/insert) for example). I'd rather stick with that.

promag commented on src/base58.cpp:
```diff
@@ -235,13 +237,64 @@ class DestinationEncoder : public boost::static_visitor<std::string>
+                }
+                return CNoDestination();
+            }
+            if (version > 16 || data.size() < 2 || data.size() > 40) return CNoDestination();
```
Return in new line?

sipa replied:
done

promag commented on src/base58.cpp:
```diff
@@ -3,15 +3,18 @@
 #include <string.h>
 #include <vector>
 #include <string>
+#include <algorithm>
```
Sort.

sipa replied:
Done.

promag commented on src/bech32.h:
```diff
@@ -0,0 +1,35 @@
+ */
+
+#include <stdint.h>
+#include <vector>
```
Sort.

sipa replied:
Done.

promag commented on src/base58.cpp:
```diff
@@ -235,13 +237,64 @@ class DestinationEncoder : public boost::static_visitor<std::string>
 {
     std::vector<unsigned char> data;
     uint160 hash;
+    auto bech = bech32::Decode(str);
```
Must come first? If not which is the cheapest?

sipa replied:
bech32 is far cheaper (no basis conversion, no SHA256).

promag commented on src/base58.cpp:
```diff
@@ -212,79 +215,102 @@ int CBase58Data::CompareTo(const CBase58Data& b58) const
-    else
+    std::vector<unsigned char> data;
+    uint160 hash;
+    auto bech = bech32::Decode(str);
```
Must come first? If not which is the cheapest?

sipa replied:
I haven't benchmarked, but Bech32 should be far cheaper (no SHA256, no basis conversion)

promag replied:
Right, in terms of functionality the order doesn't matter. But at the moment most addresses  (don't know numbers) are base58 so for now move `bech32::Decode()` after `DecodeBase58Check()`?

It would be cool to move this out of `base58.cpp`, follow up maybe?

sipa replied:
> Right, in terms of functionality the order doesn't matter. But at the moment most addresses (don't know numbers) are base58 so for now move bech32::Decode() after DecodeBase58Check()?

I was using a fail-fast approach, making the thing that most quickly fails first. You're right that as long as there are hardly any bech32 addresses, putting Base58 would be overall faster. But none of this is performance critical anyway...

promag reviewed (commented):
Fast code review. Move 4bc71a6  to #11117??

sipa commented on src/chainparams.cpp:
```diff
@@ -330,6 +334,8 @@ class CRegTestParams : public CChainParams {
         base58Prefixes[EXT_PUBLIC_KEY] = {0x04, 0x35, 0x87, 0xCF};
         base58Prefixes[EXT_SECRET_KEY] = {0x04, 0x35, 0x83, 0x94};
+
+        bech32_hrp = "rt";
```
I arbitrarily chose a bech32 prefix for regtest. Feel free to bikeshed (it doesn't even need to be just 2 characters).

promag commented on src/bech32.cpp:
```diff
@@ -0,0 +1,140 @@
+    bool ok = true;
+    for (size_t i = 0; ok && i < str.size(); ++i) {
+        unsigned char c = str[i];
+        if (c < 33 || c > 126) ok = false;
```
Early return?

sipa replied:
Done.

promag commented on src/bech32.cpp:
```diff
@@ -0,0 +1,140 @@
+        if (c >= 'a' && c <= 'z') lower = true;
+        if (c >= 'A' && c <= 'Z') upper = true;
+    }
+    if (lower && upper) ok = false;
```
Early return?

sipa replied:
Done.

promag commented on src/bech32.cpp:
```diff
@@ -0,0 +1,140 @@
+        values.resize(str.size() - 1 - pos);
+        for (size_t i = 0; i < str.size() - 1 - pos; ++i) {
+            unsigned char c = str[i + pos + 1];
+            if (CHARSET_REV[c] == -1) ok = false;
```
Early return?

sipa replied:
Done.

sipa commented on src/bech32.h:
```diff
@@ -0,0 +1,18 @@
+#include <string>
+#include <vector>
+
+namespace bech32
```
Done.

sipa replied:
Done again - I somehow lost this change.

achow101 replied:
I don't see the comment... I think you lost it again.

sipa replied:
Done.

sipa commented on src/test/base58_tests.cpp:
```diff
@@ -218,14 +247,39 @@ BOOST_AUTO_TEST_CASE(base58_keys_valid_gen)
+            BOOST_CHECK_MESSAGE(dest == dest2, strprintf("mismatch in encoding: %s", strTest));
+
+            for (char& c : address) {
+                if (c >= 'a' && c <= 'z') {
```
Unfortunately, toupper is locale-dependent, so it can't be used for consistent behaviour.

jimpo replied:
`toupper(c, std::locale::classic)`?

sipa replied:
Seems overkill.

instagibbs commented:
Include a test with v1+ address?

sipa commented:
@instagibbs Done

instagibbs commented:
super surprising to me that regtest has a different human readable part: `bcrt1`.

Legacy addresses shared between testnet and regtest. Perhaps note this somewhere in the code and/or in this PR description.

laanwj commented:
> super surprising to me that regtest has a different human readable part: bcrt1.

That was surprising to me too, but I found it a pleasant surprise. It's not like prefixes are a very rare resource, this decreases ambiguity, and fits in with regtest getting its own RPC port in #10825.

instagibbs commented:
Definitely not arguing against it!

On Wed, Sep 6, 2017 at 5:58 PM, Wladimir J. van der Laan <
notifications@github.com> wrote:

> super surprising to me that regtest has a different human readable part:
> bcrt1.
>
> That was surprising to me too, but I found it a pleasant surprise. It's
> not like prefixes are a very rare resource, this decreases ambiguity, and
> fits in with regtest getting its own RPC port in #10825
> <https://github.com/bitcoin/bitcoin/pull/10825>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/11167#issuecomment-327623706>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFgC0yE1TXYCzEvLBbEeCPST4eRCXuW4ks5sfxWYgaJpZM4PDot9>
> .
>

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1241,13 +1254,31 @@ UniValue addwitnessaddress(const JSONRPCRequest& request)
 
-    pwallet->SetAddressBook(w.result, "", "receive");
+    CScript redeemscript = GetScriptForDestination(w.result);
+    if (!w.already_witness) pwallet->AddCScript(redeemscript);
```
move this inside of the else statement below

sipa replied:
Done.

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1241,13 +1254,31 @@ UniValue addwitnessaddress(const JSONRPCRequest& request)
     }
 
-    pwallet->SetAddressBook(w.result, "", "receive");
+    CScript redeemscript = GetScriptForDestination(w.result);
```
this is more aptly described as a witness program script, since `w.result`  will always be a witness program.

sipa replied:
Fixed.

theuni commented on src/base58.cpp:
```diff
@@ -212,86 +213,51 @@ int CBase58Data::CompareTo(const CBase58Data& b58) const
+        // Base58 decoding
+        const std::vector<unsigned char>& pubkey_prefix = params.Base58Prefix(CChainParams::PUBKEY_ADDRESS);
+        if (data.size() == 20 + pubkey_prefix.size() && std::equal(pubkey_prefix.begin(), pubkey_prefix.end(), data.begin())) {
+            memcpy(hash.begin(), &data[1], 20);
```
Why is this not ```&data[pubkey_prefix.size()]```  (or ```data.data() +  pubkey_prefix.size()```) ?

sipa replied:
Nice catch, fixed!

theuni commented:
utACK other than those two nits, and excluding the bech32 implementation itself. It looks sane, but I'd like to audit more thoroughly before ACKing.

meshcollider reviewed (commented):
utACK https://github.com/bitcoin/bitcoin/pull/11167/commits/efceb6ec12321d698203b26be1a4cdeed49514b7 ... https://github.com/bitcoin/bitcoin/pull/11167/commits/8a42338bb968c213e3837e5b5c008355ce05cc23

I'm not really familiar with the technicalities of how Bech32 encoding and decoding work, but the code is consistent with what's in BIP 173. Didn't review the tests very thoroughly but they look ok

sipa commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1213,17 +1225,18 @@ UniValue addwitnessaddress(const JSONRPCRequest& request)
 
             "\nResult:\n"
-            "\"witnessaddress\",  (string) The value of the new address (P2SH of witness script).\n"
+            "\"witnessaddress\",  (string) The value of the new address (P2SH or BIP173).\n"
```
Bech32 is the generic encoding format (and should be contrasted with Base58Check), BIP173 specifies a specific way of encoding SegWit addresses using Bech32 (like P2SH/BIP13 specifies a specific way of encoding BIP16 addresses using Base58Check).

sipa commented on src/script/standard.cpp:
```diff
@@ -198,6 +203,23 @@ bool ExtractDestination(const CScript& scriptPubKey, CTxDestination& addressRet)
         return true;
+    } else if (whichType == TX_WITNESS_V0_KEYHASH) {
+        WitnessV0KeyHash hash;
+        memcpy(hash.begin(), &vSolutions[0][0], 20);
```
Fixed.

sipa commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1213,17 +1225,18 @@ UniValue addwitnessaddress(const JSONRPCRequest& request)
 
             "\nArguments:\n"
             "1. \"address\"       (string, required) An address known to the wallet\n"
+            "2. p2sh            (bool, optional) Embed inside P2SH (default: true)\n"
```
Fixed.

sipa commented on src/bech32.cpp:
```diff
@@ -0,0 +1,138 @@
+    for (size_t i = 0; i < pos; ++i) {
+        hrp += LowerCase(str[i]);
+    }
+    if (!VerifyChecksum(hrp, values)) return std::make_pair(std::string(), data());
```
Fixed.

sipa commented on src/bech32.cpp:
```diff
@@ -0,0 +1,138 @@
+        ret += CHARSET[combined[i]];
+    }
+    return ret;
+}
```
Fixed.

sipa commented:
I think I've addressed all comments. I've also added a few extra comments in some places.

> quick nit reminder, TX_WITNESS_UNKNOWN isn't explicitly handled in CombineSignatures/SignStep

Fixed.

sipa commented:
Rebased on top of #11297 to hopefully address the Travis issue.

NicolasDorier commented:
should fix https://github.com/bitcoin/bitcoin/issues/10924 I made tests on https://github.com/NicolasDorier/bitcoin/commits/importaddresssegwit can you include them @sipa  ?

NicolasDorier committed:
[RPC] Wallet: test importing of native witness scripts

Integration into segwit.py test by Pieter Wuille.

sipa commented:
@NicolasDorier Cherry-picked with some refactoring and merging into segwit.py.

NicolasDorier commented:
seems good for me.

Will try later this PR against NTumbleBit. I am using P2SH-P2WSH right now because core was not supporting otherwise. I should be able to move to pure P2WSH thanks to this PR.

ecdsa commented:
FYI, this BIP is supported by Electrum, and will be part of the next release

promag commented on src/base58.cpp:
```diff
@@ -212,86 +216,103 @@ int CBase58Data::CompareTo(const CBase58Data& b58) const
+    std::string operator()(const WitnessUnknown& id) const
+    {
+        if (id.version < 1 || id.version > 16 || id.length < 2 || id.length > 40) {
+            return "";
```
Here and in other places where an empty string is returned prefer:
```cpp
return {};
```
See https://stackoverflow.com/a/26588207.

sipa replied:
Fixed.

danra replied:
@sipa Should be rebased onto 1e46ebdf8618e585568ffc1b093c79cc9be07b57

promag commented on src/base58.cpp:
```diff
@@ -212,86 +216,103 @@ int CBase58Data::CompareTo(const CBase58Data& b58) const
+        }
+    }
+    auto bech = bech32::Decode(str);
+    if (bech.first == params.Bech32HRP() && bech.second.size() > 0) {
```
Swap conditions?

sipa replied:
Done.

promag commented on src/bech32.cpp:
```diff
@@ -0,0 +1,140 @@
+    return ret;
+}
+
+}
```
```cpp
} // namespace
```

MeshCollider replied:
Ohh heh I must have misclicked and put my comment on the wrong line, I thought I pointed this one out already but it seems I just commented on the end of a function

sipa replied:
Done.

promag commented on test/functional/segwit.py:
```diff
@@ -570,6 +584,30 @@ def run_test(self):
+            rawtxfund = self.nodes[1].signrawtransaction(rawtxfund)["hex"]
+            txid = self.nodes[1].decoderawtransaction(rawtxfund)["txid"]
+
+            self.nodes[1].sendrawtransaction(rawtxfund)
```
```py
txid = self.nodes[1].sendrawtransaction(rawtxfund)
```
And remove above?

instagibbs commented:
tACK https://github.com/bitcoin/bitcoin/pull/11167/commits/0571b9271b27ce2767e9c7034c23facb495fab65

Did a couple commits on top to get hdkey info returned with native segwit keyhashes. Something like this can be done as a quick followup, perhaps a modified version of #11089: https://github.com/instagibbs/bitcoin/commits/fullseg

achow101 commented on src/bech32.cpp:
```diff
@@ -0,0 +1,140 @@
+// Copyright (c) 2017 Pieter Wuille
```
Is this the right copyright header (here and in `bech32.h`)?

sipa replied:
I think so.

achow101 replied:
Ok, just checking since most files in this repo have `Copyright (c) The Bitcoin Core developers`

achow101 commented on src/script/ismine.cpp:
```diff
@@ -142,6 +142,9 @@ isminetype IsMine(const CKeyStore &keystore, const CScript& scriptPubKey, bool&
         break;
     }
+
+    case TX_WITNESS_UNKNOWN:
```
I think this should be above with the `TX_NULL_DATA` and `TX_NONSTANDARD` cases.

sipa replied:
Nice catch, done.

achow101 reviewed (commented):
Concept ACK. I just did a brief pass through of things that jumped out to me. Will do more review and testing later.

sdaftuar reviewed (commented):
I've only code-reviewed the Bech32 implementation (bcd3f74b649da1482c242371ddebeaf78d8b7813 ) so far, which looks good.  Left a few comment suggestions that would have been helpful for my review; feel free to take or leave, but I figured I would post them at least to help others who may be reviewing as well.

(Will continue to review the rest of the PR.)

sipa commented on src/bech32.cpp:
```diff
@@ -0,0 +1,140 @@
+}
+
+/** Find the polynomial with value coefficients mod the generator as 30-bit. */
+uint32_t PolyMod(const data& values)
```
I've added a long explanation.

sipa commented on src/bech32.cpp:
```diff
@@ -0,0 +1,140 @@
+    uint32_t chk = 1;
+    for (size_t i = 0; i < values.size(); ++i) {
+        uint8_t top = chk >> 25;
+        chk = (chk & 0x1ffffff) << 5 ^ values[i] ^
```
Done.

sipa commented on src/bech32.cpp:
```diff
@@ -0,0 +1,140 @@
+            (-((top >> 1) & 1) & 0x26508e6dUL) ^
+            (-((top >> 2) & 1) & 0x1ea119faUL) ^
+            (-((top >> 3) & 1) & 0x3d4233ddUL) ^
+            (-((top >> 4) & 1) & 0x2a1462b3UL);
```
I've replaced it with much more readable conditions. I've verified that at least GCC 6.3 compiles it to efficient code (without any jumps).

sipa commented on src/bech32.cpp:
```diff
@@ -0,0 +1,140 @@
+{
+    data enc = Cat(ExpandHRP(hrp), values);
+    enc.resize(enc.size() + 6);
+    uint32_t mod = PolyMod(enc) ^ 1;
```
Done.

gmaxwell commented on src/test/base58_tests.cpp:
```diff
@@ -218,14 +247,39 @@ BOOST_AUTO_TEST_CASE(base58_keys_valid_gen)
+                WitnessUnknown unk;
+                memcpy(unk.program, exp_payload.data(), exp_payload.size());
+                unk.length = exp_payload.size();
+                unk.version = exp_addrType[4] - 'a';
```
gross.  Can these tests be changed to work with whole script pubkeys rather than tweaking the addrType?

sipa replied:
> gross

I agree. 

> Can these tests be changed to work with whole script pubkeys rather than tweaking the addrType?

Done. I've added an extra commit to first refactor the tests into testing scriptPubKeys.

sipa committed:
Convert base58_tests from type/payload to scriptPubKey comparison

gmaxwell commented on src/test/data/base58_keys_valid.json:
```diff
@@ -422,5 +422,59 @@
+            "isTestnet": true,
+            "tryCaseFlip": true
+        }
     ]
```
no bcrt tests?

sipa replied:
Fixed.

gmaxwell reviewed (approved):
utACK.

sipa committed:
Add regtest testing to base58_tests

sipa commented:
Addressed @gmaxwell's comment about `bcrt` tests, by adding an extra commit that introduces regtest examples to base58_tests, and then extending those in the BIP173 tests commit.

jonasschnelli committed:
[Qt] tolerate BIP173/bech32 addresses during input validation

This eases the during-type validation to allow Bech32 chars.
Once the focus has been lost, the address will be properly verified through IsValidDestinationString

jonasschnelli commented:
Maybe cherry-pick  https://github.com/jonasschnelli/bitcoin/commit/f86c0585dccd29dca63cef3e6a0d434f49abac49 to enable sending to BIP173 addresses through the GUI

danra commented:
@sipa (More of a BIP 173 question) - what's the reasoning for xoring 1 to the checksum both when creating and validating it?

More specifically: `polymod` in `bech32_create_checksum()` is xored with 1 and `bech32_verify_checksum()` checks for `bench32_polymod(...) == 1`. Instead, one could avoid the xor, and check  `bench32_polymod(...) == 0`.

danra commented on src/base58.cpp:
```diff
@@ -212,86 +213,51 @@ int CBase58Data::CompareTo(const CBase58Data& b58) const
+        // Base58Check decoding
+        const std::vector<unsigned char>& pubkey_prefix = params.Base58Prefix(CChainParams::PUBKEY_ADDRESS);
+        if (data.size() == 20 + pubkey_prefix.size() && std::equal(pubkey_prefix.begin(), pubkey_prefix.end(), data.begin())) {
+            memcpy(hash.begin(), &data[pubkey_prefix.size()], 20);
```
- Use `std::copy[_n]` instead of `memcpy`
- Suggest DRYing constant lengths 20, 32. Either defining them here or using existing definition from another header.

sipa replied:
Both done, in multiple places.

sipa commented on src/base58.cpp:
```diff
@@ -5,16 +5,17 @@
-#include <vector>
-#include <string>
+
+#include <algorithm>
```
We don't have strict rules about header include order, and I'm not convinced they would be a good trade-off (more work for reviewers, only marginally more readable code).

danra replied:
IMHO Having the headers not arranged is just a more overhead to the reader. When it's a few headers, it's just a bit of overhead. When it's plenty, the overhead grows.

In any case, this specific comment is just about grouping standard header together, sorting them alphabetically, a guideline which I got the impression is mostly observed in the code.

sipa replied:
They're already sorted alphabetically and grouped together (but with C headers separate for C++ headers), which seems perfectly reasonable to me.

sipa commented on src/base58.cpp:
```diff
@@ -212,86 +213,51 @@ int CBase58Data::CompareTo(const CBase58Data& b58) const
- * The data vector contains RIPEMD160(SHA256(pubkey)), where pubkey is the serialized public key.
- * Script-hash-addresses have version 5 (or 196 testnet).
- * The data vector contains RIPEMD160(SHA256(cscript)), where cscript is the serialized redemption script.
- */
```
I've merged it into the implementation.

danra replied:
I thought I searched for the contents and didn't find it, I guess I missed it. Thanks!

sipa replied:
I only did it after you commented.

danra replied:
Ah, I misunderstood. Thanks!

sipa commented on src/base58.cpp:
```diff
@@ -328,22 +294,20 @@ bool CBitcoinSecret::SetString(const std::string& strSecret)
 {
-    return CBitcoinAddress(str).IsValid();
+    return IsValidDestination(DecodeDestination(str, Params()));
 }
```
Done.

sipa commented:
@jonasschnelli Cool, cherry-picked your commit.

@danra The xor 1 to verify the checksum is actually explained in this PR:

```c++
    // PolyMod computes what value to xor into the final values to make the checksum 0. However,
    // if we required that the checksum was 0, it would be the case that appending a 0 to a valid
    // list of values would result in a new valid list. For that reason, Bech32 requires the
    // resulting checksum to be 1 instead.
```

I'll update the BIP to explain it there as well.

danra commented on src/base58.cpp:
```diff
@@ -5,16 +5,17 @@
+
+#include <algorithm>
 #include <boost/variant/apply_visitor.hpp>
 #include <boost/variant/static_visitor.hpp>
```
The reasoning for this is the same as why user includes should come before system includes - avoid hidden dependencies.

While boost is canonical and well-tested, it's probably still better practice to have non-standard system headers be included before standard system headers, if only for consistency.

sipa replied:
Interesting, I hadn't heard about this reason. I'll take it into account in the future.

sipa commented:
Rebased on top of #11116, and fixed a bug that it discovered (`Solver` would treat non-length 20/32 v0 witness outputs as WitnessUnknown rather than failing). Thanks, @jimpo!

gmaxwell reviewed (approved):
ACK

instagibbs commented on src/qt/bitcoinaddressvalidator.cpp:
```diff
@@ -67,7 +67,7 @@ QValidator::State BitcoinAddressEntryValidator::validate(QString &input, int &po
             (ch >= 'a' && ch<='z') ||
             (ch >= 'A' && ch<='Z')) &&
-            ch != 'l' && ch != 'I' && ch != '0' && ch != 'O')
+            ch != 'I' && ch != 'O')
```
nit: I think this deserves an in-line comment.

sipa replied:
Fixed.

instagibbs commented on src/base58.cpp:
```diff
@@ -259,6 +285,34 @@ CTxDestination DecodeDestination(const std::string& str, const CChainParams& par
+    if (bech.second.size() > 0 && bech.first == params.Bech32HRP()) {
+        // Bech32 decoding
+        int version = bech.second[0];
+        if (ConvertBits<5, 8, false>(data, bech.second.begin() + 1, bech.second.end())) {
```
Not obvious to me here why it's `bech.second.begin() + 1`.

sipa replied:
Added a comment.

instagibbs commented on src/script/standard.cpp:
```diff
@@ -175,6 +182,7 @@ bool Solver(const CScript& scriptPubKey, txnouttype& typeRet, std::vector<std::v
 
 bool ExtractDestination(const CScript& scriptPubKey, CTxDestination& addressRet)
 {
+    valtype witness_program;
```
unused?

sipa replied:
Removed.

instagibbs reviewed (commented):
re-ACK, with some small nits

promag commented on src/base58.cpp:
```diff
@@ -212,86 +215,108 @@ int CBase58Data::CompareTo(const CBase58Data& b58) const
-} // namespace
+    std::string operator()(const CKeyID& id) const
+    {
+        std::vector<unsigned char> data = m_params.Base58Prefix(CChainParams::PUBKEY_ADDRESS);
```
Call `data.reserve(prefix.size() + id.size())` before filling data to avoid reallocation?

Otherwise use copy constructor? Same below.

sipa replied:
The copy constructor is used here when initializing data.

promag commented on src/bech32.cpp:
```diff
@@ -0,0 +1,191 @@
+/** Create a checksum. */
+data CreateChecksum(const std::string& hrp, const data& values)
+{
+    data enc = Cat(ExpandHRP(hrp), values);
```
Avoid copy?

sipa replied:
No copy occurs here. The temporary returned by `ExpandHRP` directly becomes the first argument to `Cat` (note that it takes a by-value argument, and not a reference, for this reason). `Cat` updates it to append `values`, and then either uses copy elision or moving to assign it to `enc`.

promag commented on src/bech32.cpp:
```diff
@@ -0,0 +1,191 @@
+data CreateChecksum(const std::string& hrp, const data& values)
+{
+    data enc = Cat(ExpandHRP(hrp), values);
+    enc.resize(enc.size() + 6); // Append 6 zeroes
```
I believe this is the 3rd possible reallocation on the same vector:
 - 1st L106: `ret.resize()`
 - 2nd L30: `x.insert(...)`
 - 3rd L130: `enc.resize(enc.size() + 6)`

Suggestion, add 2nd argument `capacity = 0` to `ExpandHRP`?

sipa replied:
Fixed. I've added a `reserve` call in `ExpandHRP` (which reserves 90 + hrp.size(), the maximum valid expanded size the checksum is computed over).

laanwj commented on src/test/base58_tests.cpp:
```diff
@@ -127,22 +84,25 @@ BOOST_AUTO_TEST_CASE(base58_keys_valid_parse)
+        bool regtest = find_value(metadata, "chain").get_str() == "regtest";
+        bool try_case_flip = find_value(metadata, "tryCaseFlip").isNull() ? false : find_value(metadata, "tryCaseFlip").get_bool();
+        if (regtest) {
+            SelectParams(CBaseChainParams::REGTEST);
```
I wonder why we don't use the SelectParams by string here.

sipa replied:
Good idea, fixed.

laanwj commented on src/test/base58_tests.cpp:
```diff
@@ -188,50 +160,37 @@ BOOST_AUTO_TEST_CASE(base58_keys_valid_gen)
-        bool isTestnet = find_value(metadata, "isTestnet").get_bool();
-        if (isTestnet)
+        bool isTestnet = find_value(metadata, "chain").get_str() == "testnet";
+        bool regtest = find_value(metadata, "chain").get_str() == "regtest";
```
Same here, `SelectParams(find_value(metadata, "chain").get_str())` would do?

sipa replied:
Fixed.

laanwj commented on src/bech32.cpp:
```diff
@@ -0,0 +1,191 @@
+const char* CHARSET = "qpzry9x8gf2tvdw0s3jn54khce6mua7l";
+
+/** The Bech32 character set for decoding. */
+const int8_t CHARSET_REV[128] = {
```
As this is indexed using a `uint8_t` I'd feel slightly more at ease if we extend this array to size 256, or add an explicit check against 128 in the loop where it is referenced. I know the check at the beginning of `Decode()` makes it impossible to hit this, but it seems something that is easily forgotten with a future code change.

sipa replied:
Fixed.

laanwj commented:
utACK +/- nits

danra commented on src/bech32.cpp:
```diff
@@ -0,0 +1,191 @@
+    // v, it corresponds to x^2 + v0*x + v1 mod g(x). As 1 mod g(x) = 1, that is the starting value
+    // for `c`.
+    uint32_t c = 1;
+    for (size_t i = 0; i < v.size(); ++i) {
```
Use a range-for loop, `i` isn't used except for accessing element (and in comments)

sipa replied:
This wasn't done originally, as the reference code is C++03, but by now it's diverged enough that I'll just bite the bullet and make it C++11 inside Core.

danra commented on src/bech32.cpp:
```diff
@@ -0,0 +1,191 @@
+/** Convert to lower case. */
+inline unsigned char LowerCase(unsigned char c)
+{
+    return (c >= 'A' && c <= 'Z') ? (c - 'A') + 'a' : c;
```
Here and later, `'A' <= c && c <= 'Z'` is better style (`c` is 'between' the values. Easier to read, harder to get wrong when used consistently)

sipa replied:
I prefer variables up front.

theuni commented on src/base58.cpp:
```diff
@@ -212,86 +215,108 @@ int CBase58Data::CompareTo(const CBase58Data& b58) const
+            std::copy(data.begin() + script_prefix.size(), data.end(), hash.begin());
+            return CScriptID(hash);
+        }
+    }
```
data.clear()

sipa replied:
Fixed.

danra replied:
@theuni Good catch!
@sipa Suggest splitting `DecodeDestination` to two functions, one for decoding Base58 and one for Bech32. This would have trivially eliminated the bug above. Another example of something which would be fixed is `hash` being used only for Base58 decoding but being visible for the entirety of the Bech32 decoding as well.

theuni replied:
+1 to this in some future follow-up.

danra commented on src/bech32.cpp:
```diff
@@ -0,0 +1,191 @@
+    // PolyMod computes what value to xor into the final values to make the checksum 0. However,
+    // if we required that the checksum was 0, it would be the case that appending a 0 to a valid
+    // list of values would result in a new valid list. For that reason, Bech32 requires the
+    // resulting checksum to be 1 instead.
```
I might be wrong, but isn't the problem with prepending a 0, rather than appending one?

sipa replied:
Thanks for taking the time to think through it, but no, this is about appending.

If the requirement was that the checksum was zero, it would mean that c(x) (the codeword polynomial) mod g(x) = 0, or in other words, c(x) is a multiple of g(x) (the generator). If c(x) is a multiple of g(x), then x*c(x) is a multiple of g(x) as well. The corresponding codeword for c(x)*x is the characters of the original with a 0 appended.

There is also a measure taken to avoid being able to prepend zeroes, namely the fact that for the codeword [a,b,c,d] we use `x^4 + a*x^3 + b*x^2 + c*x + d`, rather than `a*x^3 + b*x^2 + c*x + d` (see the comment above, about initializing the `c` variable to 1).

danra commented on src/bech32.cpp:
```diff
@@ -0,0 +1,191 @@
+    enc.resize(enc.size() + 6); // Append 6 zeroes
+    uint32_t mod = PolyMod(enc) ^ 1; // Determine what to XOR into those 6 zeroes.
+    data ret;
+    ret.resize(6);
```
Just `data ret(6);`

sipa replied:
Done.

danra commented on src/bech32.cpp:
```diff
@@ -0,0 +1,191 @@
+    data combined = Cat(values, checksum);
+    std::string ret = hrp + '1';
+    ret.reserve(ret.size() + combined.size());
+    for (size_t i = 0; i < combined.size(); ++i) {
```
Use range-for loop

sipa replied:
Done.

danra commented on src/bech32.cpp:
```diff
@@ -0,0 +1,191 @@
+        return std::make_pair(std::string(), data());
+    }
+    data values;
+    values.resize(str.size() - 1 - pos);
```
Just `data values(str.size() - 1 - pos);`

sipa replied:
Done.

danra commented on src/bech32.cpp:
```diff
@@ -0,0 +1,191 @@
+    bool lower = false, upper = false;
+    for (size_t i = 0; i < str.size(); ++i) {
+        unsigned char c = str[i];
+        if (c < 33 || c > 126) return std::make_pair(std::string(), data());
```
Here and later, just `return {}`

sipa replied:
Done.

danra commented on src/bech32.cpp:
```diff
@@ -0,0 +1,191 @@
+        if (CHARSET_REV[c] == -1) {
+            return std::make_pair(std::string(), data());
+        }
+        values[i] = CHARSET_REV[c];
```
Access `CHARSET_REV` just once, store `const int8_t& val = CHARSET_REV[c]`

sipa replied:
Fixed.

danra commented on src/bech32.cpp:
```diff
@@ -0,0 +1,191 @@
+    if (!VerifyChecksum(hrp, values)) {
+        return std::make_pair(std::string(), data());
+    }
+    return std::make_pair(hrp, data(values.begin(), values.end() - 6));
```
Here too `std::make_pair` can be removed in favor of `{}`

sipa replied:
Cool, done.

danra commented on src/test/bech32_tests.cpp:
```diff
@@ -0,0 +1,61 @@
+        if (c1 != c2) return false;
+    }
+    return true;
+}
```
Suggest using `std::tolower()` and standard algorithms to simplify the implementation

sipa replied:
Unfortunately, `tolower` is locale-dependent. I don't feel that the complexity of avoiding that is worth it for such a trivial algorithm.

sipa replied:
Locales scare me. As a reviewer, I would be far less confident that that code is correct due to it as opposed to an absolutely trivial algorithm that obvious does what it should.

theuni replied:
Agree. Seeing "locale" always means a non-trivial amount of googling is in store.

danra replied:
Ok, makes sense. In that case I adding a utility function for 'english' [is|to][lower|upper], there's too much repetition of the same (simple) functionality.

laanwj replied:
Yes, that really doesn't help. Please don't introduce dependence on locale. We prefer deterministic, well-defined, nation-independent string processing, even if that means writing some extra code at least we **know** what it's doing then. Also this is specific to bechs32 (part of the standard) so it's better to keep it self-contained.

danra replied:
@laanwj Just to repeat an earlier comment, `<locale>` has overloads of these functions where you can specify the locale, so there are no hidden dependencies on the default system locale. I do agree it might require more googling, and is perhaps a bit less obvious.

danra commented on src/test/base58_tests.cpp:
```diff
@@ -127,22 +84,25 @@ BOOST_AUTO_TEST_CASE(base58_keys_valid_parse)
-        bool isTestnet = find_value(metadata, "isTestnet").get_bool();
-        if (isTestnet)
+        bool isTestnet = find_value(metadata, "chain").get_str() == "testnet";
+        bool regtest = find_value(metadata, "chain").get_str() == "regtest";
```
`isRegtest` for consistency

sipa replied:
Already removed in response to @laanwj's comment above.

danra commented on src/base58.cpp:
```diff
@@ -212,86 +215,108 @@ int CBase58Data::CompareTo(const CBase58Data& b58) const
+        std::vector<unsigned char> data = {0};
+        ConvertBits<8, 5, true>(data, id.begin(), id.end());
+        return bech32::Encode(m_params.Bech32HRP(), data);
+    }
```
`WitnessV0KeyHash` and `WitnessV0ScriptHash` overloads have identical implementation, refactor it to a private template (templated on `id` type) function and call it from these two?

sipa replied:
I think being explicit here is clearer. I also don't expect that this correspondence will be maintained in future witness versions.

danra commented on src/base58.cpp:
```diff
@@ -212,86 +215,108 @@ int CBase58Data::CompareTo(const CBase58Data& b58) const
+                if (data.size() == scriptid.size()) {
+                    std::copy(data.begin(), data.end(), scriptid.begin());
+                    return scriptid;
+                }
```
Braces around each of `keyid` and `scriptid` segments would reduce their visibility to just the scope where they're needed.

sipa replied:
Done.

danra commented on src/base58.cpp:
```diff
@@ -212,86 +215,108 @@ int CBase58Data::CompareTo(const CBase58Data& b58) const
+            }
+            WitnessUnknown unk;
+            unk.version = version;
+            memcpy(unk.program, data.data(), data.size());
```
`std::copy`

sipa replied:
Done.

danra commented on src/base58.cpp:
```diff
@@ -212,86 +215,108 @@ int CBase58Data::CompareTo(const CBase58Data& b58) const
+        // Bech32 decoding
+        int version = bech.second[0]; // The first 5 bit symbol is the witness version (0-16)
+        // The rest of the symbols are converted witness program bytes.
+        if (ConvertBits<5, 8, false>(data, bech.second.begin() + 1, bech.second.end())) {
```
Early return if the condition fails would reduce indentation

sipa replied:
It would, but I believe it would reduce readability: there are a number of alternatives being tried, and each has its branch to process it. If it were unindented, it would need to be reindented anyway if more cases were added.

danra replied:
I agree with the principle. But, the only line I suggest dedenting is
`if (ConvertBits<5, 8, false>(data, bech.second.begin() + 1, bech.second.end()))`
which is *after* a case has been decided (Bech32 decoding), and if this condition fails, the decoding fails, so early return makes sense. It's unlikely any additional cases would be added in case of failure here.

danra commented on src/base58.cpp:
```diff
@@ -212,86 +215,108 @@ int CBase58Data::CompareTo(const CBase58Data& b58) const
+        // Bech32 decoding
+        int version = bech.second[0]; // The first 5 bit symbol is the witness version (0-16)
+        // The rest of the symbols are converted witness program bytes.
+        if (ConvertBits<5, 8, false>(data, bech.second.begin() + 1, bech.second.end())) {
```
Looks like there's no enforcing of the following rule from the BIP regarding decoding:

> Re-arrange those bits into groups of 8 bits. **Any incomplete group at the end MUST be 4 bits or less, MUST be all zeroes**, and is discarded.

danra commented on src/script/standard.cpp:
```diff
@@ -198,6 +205,23 @@ bool ExtractDestination(const CScript& scriptPubKey, CTxDestination& addressRet)
         return true;
+    } else if (whichType == TX_WITNESS_V0_KEYHASH) {
+        WitnessV0KeyHash hash;
+        memcpy(hash.begin(), vSolutions[0].data(), 20);
```
`std::copy`

sipa replied:
Done.

danra commented on src/script/standard.cpp:
```diff
@@ -198,6 +205,23 @@ bool ExtractDestination(const CScript& scriptPubKey, CTxDestination& addressRet)
+        memcpy(unk.program, vSolutions[1].data(), vSolutions[1].size());
+        unk.length = vSolutions[1].size();
+        addressRet = unk;
+        return true;
```
The header comment for `ExtractDestination` is out of date now that P2WPKH and P2WSH are supported, it says

> Currently only works for P2PK, P2PKH, and P2SH scripts.

sipa replied:
Fixed.

danra commented on src/script/standard.h:
```diff
@@ -72,14 +73,44 @@ class CNoDestination {
+    friend bool operator==(const WitnessUnknown& w1, const WitnessUnknown& w2) {
+        if (w1.version != w2.version) return false;
+        if (w1.length != w2.length) return false;
+        if (memcmp(w1.program, w2.program, w1.length)) return false;
```
`std::equal`

sipa replied:
Done.

danra commented on src/script/standard.h:
```diff
@@ -72,14 +73,44 @@ class CNoDestination {
+        if (w1.version > w2.version) return false;
+        if (w1.length < w2.length) return true;
+        if (w1.length > w2.length) return false;
+        if (memcmp(w1.program, w2.program, w1.length) < 0) return true;
```
`return std::lexicographical_compare(...);`

sipa replied:
Done.

musabi commented:
Resolved?

On Thu, Sep 28, 2017 at 12:07 PM, Wladimir J. van der Laan <
notifications@github.com> wrote:

> *@laanwj* commented on this pull request.
> ------------------------------
>
> In src/test/base58_tests.cpp
> <https://github.com/bitcoin/bitcoin/pull/11167#discussion_r141588515>:
>
> > @@ -188,50 +160,37 @@ BOOST_AUTO_TEST_CASE(base58_keys_valid_gen)
>          std::vector<unsigned char> exp_payload = ParseHex(test[1].get_str());
>          const UniValue &metadata = test[2].get_obj();
>          bool isPrivkey = find_value(metadata, "isPrivkey").get_bool();
> -        bool isTestnet = find_value(metadata, "isTestnet").get_bool();
> -        if (isTestnet)
> +        bool isTestnet = find_value(metadata, "chain").get_str() == "testnet";
> +        bool regtest = find_value(metadata, "chain").get_str() == "regtest";
>
> Same here, SelectParams(find_value(metadata, "chain").get_str()) would do?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/11167#pullrequestreview-65823840>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AdBhRyBBC4C-vDSFk4z6vB1n8MupJTWLks5sm33xgaJpZM4PDot9>
> .
>

sipa commented on src/bech32.cpp:
```diff
@@ -0,0 +1,191 @@
+typedef std::vector<uint8_t> data;
+
+/** The Bech32 character set for encoding. */
+const char* CHARSET = "qpzry9x8gf2tvdw0s3jn54khce6mua7l";
```
I think the implementation should focus on explaining why it is a correct implementation of the BIP. It doesn't need to include things the BIP already explains well - that's just gratuitously splitting the knowledge over multiple places.

Yes, the code for that optimization is in my repository, but it's very ad-hoc and not really intended for publishing.

danra replied:
Make sense regarding the information location. Where in your repo can one find the optimization code you used?

sipa replied:
https://github.com/sipa/ezbase32/blob/master/dist32.cpp

sipa commented on src/bech32.cpp:
```diff
@@ -0,0 +1,191 @@
+     1,  0,  3, 16, 11, 28, 12, 14,  6,  4,  2, -1, -1, -1, -1, -1,
+    -1, 29, -1, 24, 13, 25,  9,  8, 23, -1, 18, 22, 31, 27, 19, -1,
+     1,  0,  3, 16, 11, 28, 12, 14,  6,  4,  2, -1, -1, -1, -1, -1
+};
```
Sounds good, but seems out of scope for this PR.

sipa commented on src/bech32.cpp:
```diff
@@ -0,0 +1,191 @@
+    // between {}. In this finite field, addition is just XOR of the corresponding numbers. For
+    // example, {27} + {13} = {27 ^ 13} = {22}. Multiplication is more complicated, and requires
+    // treating the bits of values themselves as coefficients of a polynomial over a smaller field,
+    // GF(2), and multiplying those polynomials mod a^5 + a^3 + 1. For example, {5} * {26} =
```
Yes and no.

There is only one field of size 32 _up to isomorphism_. However, the multiplication of GF(32) elements _when represented as bits_ absolutely depends on the choice of the polynomial - and this is the one used in Bech32. It just so happens that you could do a mapping of the input values to the respective elements in an isomorphic representation (which uses another polynomial), do the multiplication, and map things back.

I think the current explanation is sufficiently accurate.

danra replied:
> However, the multiplication of GF(32) elements when represented as bits absolutely depends on the choice of the polynomial - and this is the one used in Bech32
Yes, that's what I meant - emphasizing the fact that this polynomial was chosen, not that it is *the* one required for multiplication. I think this point could be better made. Though, I agree that even without this nit, the documentation is great :)

sipa replied:
Well I think it is _the_ one required for multiplication.

The specific representation of GF(32) we use, is defined as bits representing polynomials over GF(2) modulo exactly the polynomial `a^5 + a^3 + 1` and no other.

Yes, other representations using other polynomials exist, and given that all size 32 fields are isomorphic, it's guaranteed that a mapping exists between them which is consistent with addition and multiplication. But we're not using any of those :)

danra replied:
> Or do you mean that I should point out that Bech32 could have used any of the other irreducible polynomials?

Yes. I just suggest a more precise phrasing. e.g., instead of
> Multiplication is more complicated, and requires [...] multiplying those polynomials mod a^5 + a^3 + 1

write 

> Multiplication is more complicated, and requires [...] multiplying those polynomials mod an irreducible polynomial, chosen as a^5 + a^3 + 1 in Bech32."

danra commented:
there’s a version of tolower where you can specify the locale yourself, so there is no concern about the wrong one being set in the system

> On 29 Sep 2017, at 2:04, Pieter Wuille <notifications@github.com> wrote:
> 
> @sipa commented on this pull request.
> 
> In src/test/bech32_tests.cpp:
> 
> > +#include <boost/test/unit_test.hpp>
> +
> +BOOST_FIXTURE_TEST_SUITE(bech32_tests, BasicTestingSetup)
> +
> +bool CaseInsensitiveEqual(const std::string &s1, const std::string &s2)
> +{
> +    if (s1.size() != s2.size()) return false;
> +    for (size_t i = 0; i < s1.size(); ++i) {
> +        char c1 = s1[i];
> +        if (c1 >= 'A' && c1 <= 'Z') c1 -= ('A' - 'a');
> +        char c2 = s2[i];
> +        if (c2 >= 'A' && c2 <= 'Z') c2 -= ('A' - 'a');
> +        if (c1 != c2) return false;
> +    }
> +    return true;
> +}
> Unfortunately, tolower is locale-dependent. I don't feel that the complexity of avoiding that is worth it for such a trivial algorithm.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

sipa commented on src/base58.cpp:
```diff
@@ -212,86 +215,108 @@ int CBase58Data::CompareTo(const CBase58Data& b58) const
+            std::copy(data.begin() + script_prefix.size(), data.end(), hash.begin());
+            return CScriptID(hash);
+        }
+    }
```
I prefer to do that later.

sipa commented:
I believe I've addressed or at least responded to all comments by @promag, @laanwj, @theuni, and @danra.

I've also updated the Bech32 tests with some new additions to the BIP.

sipa commented on test/functional/segwit.py:
```diff
@@ -570,6 +584,29 @@ def run_test(self):
+
+            # Assert it is properly saved
+            self.stop_node(1)
+            self.start_node(1)
```
```python
  File "./segwit.py", line 606, in run_test
    self.restart_node(1)
AttributeError: 'SegWitTest' object has no attribute 'restart_node'
```

laanwj commented:
The commentary seems to have reached the level of nano-nits, which is likely an indication this is ready for merge.

TheBlueMatt reviewed (commented):
postumous Matt was here and didn't finish review.

theuni commented on src/bech32.cpp:
```diff
@@ -0,0 +1,191 @@
+    ret.reserve(hrp.size() + 90);
+    ret.resize(hrp.size() * 2 + 1);
+    for (size_t i = 0; i < hrp.size(); ++i) {
+        unsigned char c = hrp[i];
```
Unlikely to matter, but shouldn't this be ```LowerCase(hrp[i])``` ?

MarcoFalke commented:
Removing from backport
